File: haskell/bob/Bob.hs
module Bob (responseFor) where

import Data.Char (isAlpha, isSpace, isUpper)
import Data.List (isSuffixOf)

responseFor :: String -> String
responseFor message
  | isSilence message  = "Fine. Be that way!"
  | isYelling message  = "Woah, chill out!"
  | isQuestion message = "Sure."
  | otherwise          = "Whatever."

isSilence :: String -> Bool
isSilence = all isSpace

isYelling :: String -> Bool
isYelling message =
  let letters = filter isAlpha message
  in not (null letters) && all isUpper letters

isQuestion :: String -> Bool
isQuestion = isSuffixOf "?"

File: javascript/bob/bob.js
module.exports = function () {
  'use strict';

  var isSilence, isYelling, isQuestion;

  isSilence = function (message) {
    return !message.match(/\S/);
  };

  isYelling = function (message) {
    return !message.match(/[a-z]/);
  };

  isQuestion = function (message) {
    return !message.match(/[^?]$/);
  };

  return {
    hey: function (message) {
      if (isSilence(message)) { return 'Fine. Be that way!'; }
      if (isYelling(message)) { return 'Woah, chill out!'; }
      if (isQuestion(message)) { return 'Sure.'; }
      return 'Whatever.';
    }
  };
};
